mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala

```diff
@@ -262,8 +264,323 @@ class LogisticRegression @Since("1.2.0") (
+             {{{
+               b = \log{P(1) / P(0)} = \log{count_1 / count_0}
+             }}}
+           */
+        initialCoefficientsWithIntercept.toArray(numFeatures) = math.log(
+          histogram(1) / histogram(0))
+      }
+
+      time = System.currentTimeMillis()
+      timer.push(time)
+
+      val states = optimizer.iterations(new CachedDiffFunction(costFun),
+        initialCoefficientsWithIntercept.asBreeze.toDenseVector)
+
+      /*
+           Note that in Logistic Regression, the objective history (loss + regularization)
+           is log-likelihood which is invariance under feature standardization. As a result,
+           the objective history from optimizer is the same as the one in the original space.
+         */
+      val arrayBuilder = mutable.ArrayBuilder.make[Double]
+      var state: optimizer.State = null
+      while (states.hasNext) {
+        state = states.next()
+        arrayBuilder += state.adjustedValue
+      }
+
+      if (state == null) {
+        val msg = s"${optimizer.getClass.getName} failed."
+        logError(msg)
+        throw new SparkException(msg)
+      }
+
+      time = System.currentTimeMillis()
+      val time_diff = time - timer.pop()
+      printf("%s] *****optimizer took %d ms regParam=%f: %s\n", uid, time - timer.pop(), $(regParam), state.convergedReason.toString());
+
+      if (time_diff > 35000) {
+        CuAccManager.dump(dataset_rdd, time_diff.toString() + ".dat")
+      }
+
+      /*
+           The coefficients are trained in the scaled space; we're converting them back to
+           the original space.
+           Note that the intercept in scaled space and original space is the same;
+           as a result, no scaling is needed.
+         */
+      val rawCoefficients = state.x.toArray.clone()
+      var i = 0
+      while (i < numFeatures) {
+        rawCoefficients(i) *= { if (featuresStd(i) != 0.0) 1.0 / featuresStd(i) else 0.0 }
+        i += 1
+      }
+
+      if ($(fitIntercept)) {
+        (Vectors.dense(rawCoefficients.dropRight(1)).compressed, rawCoefficients.last,
+          arrayBuilder.result())
+      } else {
+        (Vectors.dense(rawCoefficients).compressed, 0.0, arrayBuilder.result())
+      }
+    }
+
+    CuAccManager.destroy(exec_rdd, exec_info)
+
+    val model = copyValues(new LogisticRegressionModel(uid, coefficients, intercept))
+    val (summaryModel, probabilityColName) = model.findSummaryModelAndProbabilityCol()
+    val logRegSummary = new BinaryLogisticRegressionTrainingSummary(
+      summaryModel.transform(dataset),
+      probabilityColName,
+      $(labelCol),
+      $(featuresCol),
+      objectiveHistory)
+    model.setSummary(logRegSummary)
+  }
+
   protected[spark] def train(dataset: Dataset[_], handlePersistence: Boolean):
       LogisticRegressionModel = {
+    val dataset_rdd = dataset.select(col($(labelCol)).cast(DoubleType), col($(featuresCol))).rdd
+    val lib_gpu = CuAccManager.get_lib_gpu(dataset_rdd.conf)
```

Comment:
This code will slow down the original non-GPU code because it creates a RDD from the DataFrame in order to determine if it should run on the GPU or not.  Is there a way to make this determination another way, and create the RDD only if the GPU path is to be taken?